File: src/hyperhandler/risk/collector.py
"""Trade result collection for circuit breaker tracking."""

from datetime import datetime, timezone
from decimal import Decimal
from typing import TYPE_CHECKING

from hyperhandler.models.order import Position
from hyperhandler.models.risk import TradeResult

if TYPE_CHECKING:
    from hyperhandler.client.info import InfoClient
    from hyperhandler.storage import Storage


class TradeResultCollector:
    """Collects closed trade results for circuit breaker tracking.

    Two collection strategies:
    1. On-close: When position is closed via CLI command
    2. Reconcile: Periodic sync from HL fills API
    """

    def __init__(self, storage: "Storage", network: str):
        self.storage = storage
        self.network = network
        self._recorded_fills: set[str] = set()

    async def collect_from_fills(
        self,
        info_client: "InfoClient",
        address: str,
        since_timestamp: datetime | None = None,
    ) -> list[TradeResult]:
        """Reconcile trade results from HL user fills.

        This matches fills to known positions and calculates PnL.
        Should be called periodically or before risk evaluation.

        Args:
            info_client: InfoClient for API calls.
            address: User's wallet address.
            since_timestamp: Only process fills after this time.

        Returns:
            List of newly recorded TradeResults.
        """
        fills = await info_client.get_user_fills(address, limit=100)

        results: list[TradeResult] = []
        for fill in fills:
            # Check if this fill closes a position
            closed_pnl = fill.get("closedPnl")
            if not closed_pnl:
                continue  # Not a closing fill

            # Generate unique fill ID
            fill_id = f"{fill.get('oid', '')}_{fill.get('time', '')}"
            if fill_id in self._recorded_fills:
                continue  # Already processed

            # Check timestamp filter
            fill_time = datetime.fromtimestamp(
                fill.get("time", 0) / 1000, tz=timezone.utc
            )
            if since_timestamp and fill_time < since_timestamp:
                continue

            # Extract entry price from startPosition if available
            start_pos = fill.get("startPosition", {})
            entry_px = start_pos.get("entryPx", fill.get("px", "0"))

            result = TradeResult(
                coin=fill.get("coin", ""),
                side="long" if fill.get("side") == "B" else "short",
                entry_price=Decimal(str(entry_px)),
                exit_price=Decimal(str(fill.get("px", "0"))),
                size=Decimal(str(fill.get("sz", "0"))),
                pnl=Decimal(str(closed_pnl)),
                fees=Decimal(str(fill.get("fee", "0"))),
                funding_paid=Decimal("0"),  # Not available in fills
                opened_at=datetime.fromtimestamp(
                    start_pos.get("time", fill.get("time", 0)) / 1000,
                    tz=timezone.utc,
                ),
                closed_at=fill_time,
            )

            # Save to storage
            result_id = self.storage.save_trade_result(result, self.network)
            result.id = result_id
            self._recorded_fills.add(fill_id)
            results.append(result)

        return results

    def record_close(
        self,
        position: Position,
        exit_price: Decimal,
        fees: Decimal,
        funding_paid: Decimal = Decimal("0"),
        signal_id: int | None = None,
    ) -> TradeResult:
        """Record a position close initiated via CLI.

        Called by ExchangeClient.close_position() or similar.

        Args:
            position: The position being closed.
            exit_price: Exit price.
            fees: Trading fees paid.
            funding_paid: Funding payments.
            signal_id: Associated signal ID if known.

        Returns:
            TradeResult with ID populated.
        """
        pnl = self._calculate_pnl(position, exit_price, fees, funding_paid)

        result = TradeResult(
            signal_id=signal_id,
            coin=position.coin,
            side="long" if position.is_long else "short",
            entry_price=position.entry_price,
            exit_price=exit_price,
            size=position.abs_size,
            pnl=pnl,
            fees=fees,
            funding_paid=funding_paid,
            opened_at=position.opened_at or datetime.now(timezone.utc),
            closed_at=datetime.now(timezone.utc),
        )

        result_id = self.storage.save_trade_result(result, self.network)
        result.id = result_id
        return result

    def get_recent_results(self, limit: int = 50) -> list[TradeResult]:
        """Get recent trade results from storage.

        Args:
            limit: Maximum results to return.

        Returns:
            List of TradeResults ordered by closed_at desc.
        """
        return self.storage.get_recent_trade_results(self.network, limit=limit)

    def _calculate_pnl(
        self,
        position: Position,
        exit_price: Decimal,
        fees: Decimal,
        funding_paid: Decimal,
    ) -> Decimal:
        """Calculate realized PnL.

        Args:
            position: Position being closed.
            exit_price: Exit price.
            fees: Trading fees.
            funding_paid: Funding payments.

        Returns:
            Net realized PnL.
        """
        if position.is_long:
            gross_pnl = (exit_price - position.entry_price) * position.abs_size
        else:
            gross_pnl = (position.entry_price - exit_price) * position.abs_size

        return gross_pnl - fees - funding_paid

File: internal/risk/collector.go
package risk

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"github.com/shopspring/decimal"

	"github.com/skilus/hyperhandler/internal/models"
)

// TradeResultStore is the storage surface the collector needs. Implemented by
// the storage layer (SPEC-007 Phase 5). Mirrors the Storage methods used by
// collector.py.
type TradeResultStore interface {
	SaveTradeResult(result models.TradeResult, network string) (int64, error)
	GetRecentTradeResults(network string, limit int) ([]models.TradeResult, error)
}

// FillsProvider is the Info API surface needed to reconcile from fills.
// Satisfied by *client.InfoClient.
type FillsProvider interface {
	GetUserFills(ctx context.Context, address string, limit int) ([]json.RawMessage, error)
}

// Collector collects closed trade results for circuit-breaker tracking. Mirrors
// collector.py:TradeResultCollector. The in-memory fill dedup mirrors the Python
// behaviour; the UNIQUE-on-fill_id upsert is scheduled for Phase 5 (B.5). The
// clock is injectable for deterministic tests.
type Collector struct {
	Storage TradeResultStore
	Network string

	recordedFills map[string]struct{}
	now           func() time.Time
}

// NewCollector builds a Collector backed by storage for a network.
func NewCollector(storage TradeResultStore, network string) *Collector {
	return &Collector{
		Storage:       storage,
		Network:       network,
		recordedFills: map[string]struct{}{},
		now:           time.Now,
	}
}

// WithClock overrides the clock (for tests).
func (c *Collector) WithClock(now func() time.Time) *Collector {
	c.now = now
	return c
}

// startPosition mirrors the nested startPosition object in a fill.
type startPosition struct {
	EntryPx *string `json:"entryPx"`
	Time    *int64  `json:"time"`
}

// rawFill mirrors the fields the collector reads from an HL user fill.
type rawFill struct {
	Coin          string         `json:"coin"`
	Px            string         `json:"px"`
	Sz            string         `json:"sz"`
	Side          string         `json:"side"`
	Time          int64          `json:"time"`
	Oid           int64          `json:"oid"`
	ClosedPnl     *string        `json:"closedPnl"`
	Fee           string         `json:"fee"`
	StartPosition *startPosition `json:"startPosition"`
}

// CollectFromFills reconciles trade results from HL user fills, matching closing
// fills to PnL. Returns the newly recorded results. Mirrors
// collector.py:collect_from_fills. A nil sinceTimestamp processes all fills.
func (c *Collector) CollectFromFills(
	ctx context.Context,
	info FillsProvider,
	address string,
	sinceTimestamp *time.Time,
) ([]models.TradeResult, error) {
	rawFills, err := info.GetUserFills(ctx, address, 100)
	if err != nil {
		return nil, err
	}

	var results []models.TradeResult
	for _, raw := range rawFills {
		var fill rawFill
		if err := json.Unmarshal(raw, &fill); err != nil {
			return nil, fmt.Errorf("decode fill: %w", err)
		}

		// Skip non-closing fills (closedPnl absent or empty).
		if fill.ClosedPnl == nil || *fill.ClosedPnl == "" {
			continue
		}

		// Dedup by unique fill ID.
		fillID := fmt.Sprintf("%d_%d", fill.Oid, fill.Time)
		if _, seen := c.recordedFills[fillID]; seen {
			continue
		}

		fillTime := time.UnixMilli(fill.Time).UTC()
		if sinceTimestamp != nil && fillTime.Before(*sinceTimestamp) {
			continue
		}

		// Entry price from startPosition if available, else the fill price.
		entryPx := fill.Px
		openedMs := fill.Time
		if fill.StartPosition != nil {
			if fill.StartPosition.EntryPx != nil {
				entryPx = *fill.StartPosition.EntryPx
			}
			if fill.StartPosition.Time != nil {
				openedMs = *fill.StartPosition.Time
			}
		}

		side := "short"
		if fill.Side == "B" {
			side = "long"
		}

		result := models.TradeResult{
			Coin:        fill.Coin,
			Side:        side,
			EntryPrice:  mustParseDec(entryPx),
			ExitPrice:   mustParseDec(fill.Px),
			Size:        mustParseDec(fill.Sz),
			Pnl:         mustParseDec(*fill.ClosedPnl),
			Fees:        mustParseDec(fill.Fee),
			FundingPaid: decimal.Zero, // not available in fills
			OpenedAt:    time.UnixMilli(openedMs).UTC(),
			ClosedAt:    fillTime,
		}

		id, err := c.Storage.SaveTradeResult(result, c.Network)
		if err != nil {
			return nil, err
		}
		result.ID = &id
		c.recordedFills[fillID] = struct{}{}
		results = append(results, result)
	}

	return results, nil
}

// RecordClose records a position close initiated via the CLI. Mirrors
// collector.py:record_close.
func (c *Collector) RecordClose(
	position models.Position,
	exitPrice, fees, fundingPaid decimal.Decimal,
	signalID *int64,
) (models.TradeResult, error) {
	pnl := c.calculatePnl(position, exitPrice, fees, fundingPaid)

	side := "short"
	if position.IsLong() {
		side = "long"
	}
	openedAt := c.now().UTC()
	if position.OpenedAt != nil {
		openedAt = *position.OpenedAt
	}

	result := models.TradeResult{
		SignalID:    signalID,
		Coin:        position.Coin,
		Side:        side,
		EntryPrice:  position.EntryPrice,
		ExitPrice:   exitPrice,
		Size:        position.AbsSize(),
		Pnl:         pnl,
		Fees:        fees,
		FundingPaid: fundingPaid,
		OpenedAt:    openedAt,
		ClosedAt:    c.now().UTC(),
	}

	id, err := c.Storage.SaveTradeResult(result, c.Network)
	if err != nil {
		return models.TradeResult{}, err
	}
	result.ID = &id
	return result, nil
}

// GetRecentResults returns recent trade results from storage. Mirrors
// collector.py:get_recent_results.
func (c *Collector) GetRecentResults(limit int) ([]models.TradeResult, error) {
	return c.Storage.GetRecentTradeResults(c.Network, limit)
}

// calculatePnl computes realized net PnL. Mirrors collector.py:_calculate_pnl.
func (c *Collector) calculatePnl(
	position models.Position,
	exitPrice, fees, fundingPaid decimal.Decimal,
) decimal.Decimal {
	var grossPnl decimal.Decimal
	if position.IsLong() {
		grossPnl = exitPrice.Sub(position.EntryPrice).Mul(position.AbsSize())
	} else {
		grossPnl = position.EntryPrice.Sub(exitPrice).Mul(position.AbsSize())
	}
	return grossPnl.Sub(fees).Sub(fundingPaid)
}

// mustParseDec parses a decimal string, treating malformed/empty input as zero
// (mirrors Python's Decimal(str(x)) over API strings, which are well-formed).
func mustParseDec(s string) decimal.Decimal {
	d, err := decimal.NewFromString(s)
	if err != nil {
		return decimal.Zero
	}
	return d
}
